### PR TITLE
chore: refactor ChainStore::genesis to inherit the genesis value from the chain config 

### LINKED ISSUE #2410: chore: refactor `ChainStore::genesis` to inherit the genesis value from the chain config
**Issue summary**

Right now, `ChainStore::genesis` has the option of being undefined. This causes a lot of `unwrap`s in other parts of the code which require the genesis block to be set.

We should take the genesis information from the chain configuration and use it to initialize the genesis key in the database. That way, `ChainStore::genesis` will always return the right value.

- [ ] Make sure the database always contains a reference to the genesis block (see `set_genesis`).
- [ ] Update the type of `ChainStore::genesis` and update all calls to this function.

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
## Summary of changes

Changes introduced in this pull request:
- 

## Reference issue to close (if applicable)

Closes #2410

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: utils/genesis/src/lib.rs [hunk 1]
```diff
@@ -103,12 +103,9 @@ where
...
 
-    let store_genesis = chain_store.genesis()?;
+    let store_genesis = chain_store.genesis();
 
-    if store_genesis
-        .map(|store| store == genesis_block)
-        .unwrap_or_default()
-    {
+    if store_genesis.is_ok() && store_genesis.unwrap() == genesis_block {
         debug!("Genesis from config matches Genesis from store");
```
Comment (lemmih) [inline utils/genesis/src/lib.rs:109]: This debug message isn't important. Let's refactor the code to always set the genesis.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: utils/genesis/src/lib.rs (hunk 1 above)
Comment (tyshko5) [reply utils/genesis/src/lib.rs]: Removed the message

File: utils/genesis/src/lib.rs [hunk 2]
```diff
@@ -103,14 +103,9 @@ where
...
-    let store_genesis = chain_store.genesis()?;
+    let store_genesis = chain_store.genesis();
 
-    if store_genesis
-        .map(|store| store == genesis_block)
-        .unwrap_or_default()
-    {
-        debug!("Genesis from config matches Genesis from store");
-    } else {
+    if store_genesis.is_err() {
```
Comment (lemmih) [inline utils/genesis/src/lib.rs:108]: You can delete `store_genesis` altogether. We always want to update the genesis in the DB.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: utils/genesis/src/lib.rs (hunk 2 above)
Comment (tyshko5) [reply utils/genesis/src/lib.rs]: Deleted

**Comment by lemmih** [comment]
Good. You still need to go through the four check boxes in the PR description.

**Comment by lemmih** [comment]
You can run forest locally to see what's going wrong.

File: blockchain/chain/src/store/chain_store.rs [hunk 3]
```diff
@@ -731,14 +731,13 @@ where
 }
 
 /// Returns the genesis block from storage.
-pub fn genesis<DB>(db: &DB) -> Result<Option<BlockHeader>, Error>
+pub fn genesis<DB>(db: &DB) -> Result<BlockHeader, Error>
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:734]: This function should only be used from the `ChainStore`. Let's move it into `ChainStore::genesis`.

Comment (lemmih) [reply blockchain/chain/src/store/chain_store.rs:734]: And let's do the same with `set_genesis`. That function should also only be used through the `ChainStore`.

File: blockchain/chain/src/store/chain_store.rs [hunk 4]
```diff
@@ -731,14 +731,13 @@ where
...
 where
     DB: Blockstore + Store,
 {
-    Ok(db
-        .read(GENESIS_KEY)?
-        .map(|bz| BlockHeader::unmarshal_cbor(&bz))
-        .transpose()?)
+    db.read(GENESIS_KEY)?
+        .map(|bz| BlockHeader::unmarshal_cbor(&bz).map_err(|e| Error::Other(e.to_string())))
+        .ok_or(Error::Other("no genesis".to_string()))?
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:740]: Maybe the error message could be `Genesis key not defined in database`

**Review by lemmih — CHANGES_REQUESTED** [review]

File: blockchain/chain/src/store/chain_store.rs (hunk 4 above)
Comment (tyshko5) [reply blockchain/chain/src/store/chain_store.rs]: Changed

File: blockchain/chain/src/store/chain_store.rs (hunk 3 above)
Comment (tyshko5) [reply blockchain/chain/src/store/chain_store.rs]: Moved

**Review by LesnyRumcajs — APPROVED** [review]

**Review by sudo-shashank — APPROVED** [review]

**Pull request merged**